pallene/reaching_definitions.lua

```diff
@@ -0,0 +1,377 @@
+
+local ir = require "pallene.ir"
+local location = require "pallene.location"
+
+--local inspect = require "inspect"
+--local ppi = function(t,f) return inspect(t,{newline='',process=f}) end
+--local pp = function(t,f) return inspect(t,{process=f}) end
+
+local RD = {}
+-- start set operations -----------
+
+local function copy(t1,t2)
+    for k,v in pairs(t2) do t1[k] = v end
```

Comment:
> This function is using pairs, while all the other ones are using ipairs. It should be one or the other.

Should be ipairs, will be fixed.

> In uninitialized.lua we represent sets as tables where the set elements are used as keys. Is there some reason why this can't be done here as well? It would avoid having to call the deduplicate function all the time.

My first implementation was using keys as elements but it was quite hard to debug. I agree that it is better, but since the sets are never too big I just postponed it as an optimization. I can convert it to the key format before merging.

Reply:
Interesting. Was it because it was harder to inspect the sets or was it because "pairs" is nondeterministic?

Reply:
both because of pairs being nondeterministic and length operator not being available for this kind of view, also printing being a bit of a hassle... 

But I could just define a internal field with the element count (updated via metamethods) and a __tostring method with 
```
str = '{'
for k, v in next, t do
  str = str..','..k
end 
return str..'}'
``` 
the effort didn't seem to compensate..

Reply:
What are the places that we currently use the number of elements in the set? I see it's being used in the `equal` method but is there another place that I missed?

I ask this because at least in the case of `equal` then I suspect that keeping track of the number of elements in the set is not necessary. We could replace `equal(x, {})` with an `isempty` function. We could also replace `equal(old, new)` by changing the code that updates the sets to set a boolean flag if it makes any changes. I think that is the only two places that we are using `equal` right now.

Reply:
Also, I wouldn't be able to iterate via ipairs

Reply:
I am planning to change the implementation to use keys as set elements in the close future